File: 1002.Find-Common-Characters.go
// https://leetcode.com/problems/find-common-characters/
//
// algorithms
// Easy (70.81%)
// Total Accepted:    4,545
// Total Submissions: 6,419
// beats 100.0% of golang submissions

package leetcode

func commonChars(A []string) []string {
	minNum := make([]int, 26)
	for _, n := range A[0] {
		minNum[int(n)-97]++
	}
	tmp := make([]int, 26)
	for _, s := range A[1:] {
		tmp = make([]int, 26)
		for _, n := range s {
			tmp[int(n)-97]++
		}
		for i := 0; i < 26; i++ {
			minNum[i] = min(minNum[i], tmp[i])
		}
	}

	var res []string

	for i, n := range minNum {
		s := string(i + 97)
		for j := 0; j < n; j++ {
			res = append(res, s)
		}
	}

	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 1002.Find-Common-Characters.java
// https://leetcode.com/problems/find-common-characters/
//
// algorithms
// Easy (66.61%)
// Total Accepted:    14,803
// Total Submissions: 22,225
// beats 95.22% of java submissions

class Solution {
    public List<String> commonChars(String[] A) {
        int ch[] = new int[26];
        Arrays.fill(ch, 101);
        
        for (String s : A) {
            int tmp[] = new int[26];
            for (char c : s.toCharArray()) {
                tmp[c - 'a']++;
            }
            getSmaller(ch, tmp);
        }
        
        List<String> res = new ArrayList<>();
        for (int i = 0; i < 26; i++) {
            String s = String.valueOf((char)(i + 'a'));
            for (int j = 0; j < ch[i]; j++) {
                res.add(s);
            }
        }
        
        return res;
    }
    
    public void getSmaller(int[] ch, int[] b) {
        for (int i = 0; i < 26; i++) {
            if (ch[i] > b[i]) {
                ch[i] = b[i];
            }
        }
    }
}